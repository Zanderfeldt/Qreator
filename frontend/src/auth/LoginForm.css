/* LogInForm.css */

.login-card {
  background-color: #ffffff62 ;
  border-radius: 10px;
  border: solid rgba(39, 35, 33, 0.192) 0.5px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.295);
  width: 320px;
  margin: 80px auto;
  padding: 20px;
  text-align: center;
  
}

.login-card h2 {
  color: #333;
  font-size: 24px;
  margin-bottom: 10px;
}


.login-card input {
  width: 80%;
  padding: 2px;
  margin: 8px 0;
  border: 1px solid #ccc;
  border-radius: 4px;
  font-size: 12px;
  height: 25px;
  font-family: 'Mooli', sans-serif;
}

.login-card button {
  background-color: #62865d;
  color: #fff;
  padding: 5px 10px;
  border: none;
  border-radius: 1px;
  cursor: pointer;
  font-size: 16px;
  margin-top: 10px;
  font-family: 'Mooli', sans-serif;
  
}

.login-card button:hover {
  background-color: #70dd5a;
}

.error-message {
  color: red; /* Set text color to red */
  font-size: 12px; /* Set font size to make it smaller */
}