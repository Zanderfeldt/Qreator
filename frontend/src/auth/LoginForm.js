import React, { useState } from 'react';
import { useHistory } from "react-router-dom";
import './LoginForm.css';

function LoginForm({login}) {

  const INITIAL_STATE = {
    username: '',
    password: '',
  }

  const [formData, setFormData] = useState(INITIAL_STATE);
  const [showAlert, setShowAlert] = useState('');
  const [errors, setErrors] = useState(INITIAL_STATE)
  const history = useHistory();

  const handleSubmit = async (e) => {
    e.preventDefault();
    // Check if any of the fields are empty, and set the error messages
    const newErrors = { ...INITIAL_STATE };
    if (formData.username.trim() === '') {
      newErrors.username = 'Username is required';
    }
    if (formData.password.trim() === '') {
      newErrors.password = 'Password is required';
    }
    setErrors(newErrors);

    // If there are any errors, don't proceed with login
    if (Object.values(newErrors).some((error) => error !== '')) {
      return;
    }

    let result = await login(formData);
    if (result.success) {
      setFormData(INITIAL_STATE);
      history.push('/');
    } else {
      setShowAlert(result.message);
    }
  };

  const handleChange = e => {
    const {name, value} = e.target;
    setFormData(fData => ({
      ...fData,
      [name]: value
    }));

    // Clear the error message when the user starts typing
    setErrors((prevErrors) => ({
      ...prevErrors,
      [name]: '',
    }));
  };

  
  return (
    <div className='login-card'>
      <h2>Log In</h2>
      <small style={{color: 'red'}}>{showAlert}</small>
      <form onSubmit={handleSubmit}>
        <input
          id='username'
          name='username'
          placeholder='Username'
          value={formData.username}
          onChange={handleChange}
          data-testid='Username'
          />
        {errors.username && <div className="error-message">{errors.username}</div>}
        <input
          type='password'
          id='password'
          name='password'
          placeholder='Password'
          value={formData.password}
          onChange={handleChange}
          data-testid='Password'
          />
          {errors.password && <div className="error-message">{errors.password}</div>}
          <button>Submit</button>
      </form>

    </div>
  )
}

export default LoginForm;