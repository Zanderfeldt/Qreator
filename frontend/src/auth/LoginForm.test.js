import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import LoginForm from './LoginForm';
import { MemoryRouter, Route } from 'react-router-dom';

const mockLogin = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: jest.fn(),
    location: { pathname: '/' },
  }),
}));

describe('LoginForm', () => {
  // Smoke Test
  it('should render without errors', () => {
    render(<LoginForm login={() => {}} />);
  });

  // Snapshot Test
  it('should match the snapshot', () => {
    const { asFragment } = render(<LoginForm login={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const renderWithRouter = () => {
    return render(
      <MemoryRouter initialEntries={['/login']}>
        <Route path="/login">
          <LoginForm login={mockLogin} />
        </Route>
      </MemoryRouter>
    );
  };

  it('displays validation errors for empty fields', () => {
    renderWithRouter();

    const submitButton = screen.getByText('Submit');
    fireEvent.click(submitButton);

    // Ensure that error messages are displayed for both fields
    expect(screen.getByText('Username is required')).toBeInTheDocument();
    expect(screen.getByText('Password is required')).toBeInTheDocument();
  });

  it('redirects to home page after successful login', async () => {
    // Mock a successful login
    mockLogin.mockResolvedValue({ success: true });

    renderWithRouter();

    // Fill out the form
    const usernameInput = screen.getByTestId('Username');
    const passwordInput = screen.getByTestId('Password');
    const submitButton = screen.getByText('Submit');

    fireEvent.change(usernameInput, { target: { value: 'testuser' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });
    fireEvent.click(submitButton);

    // Wait for the redirection to occur
    await waitFor(() => {
      expect(window.location.pathname).toBe('/');
    });
  });
});
