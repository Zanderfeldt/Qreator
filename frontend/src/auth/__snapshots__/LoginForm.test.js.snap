// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="login-card"
  >
    <h2>
      Log In
    </h2>
    <small
      style="color: red;"
    />
    <form>
      <input
        data-testid="Username"
        id="username"
        name="username"
        placeholder="Username"
        value=""
      />
      <input
        data-testid="Password"
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
